Sources/AWSLambdaRuntimeCore/LambdaHandler.swift

```diff
@@ -129,18 +129,42 @@ public protocol EventLoopLambdaHandler: ByteBufferLambdaHandler {
 public extension EventLoopLambdaHandler {
     /// Driver for `ByteBuffer` -> `In` decoding and `Out` -> `ByteBuffer` encoding
     func handle(context: Lambda.Context, event: ByteBuffer) -> EventLoopFuture<ByteBuffer?> {
-        switch self.decodeIn(buffer: event) {
+        let segment = context.tracer.beginSegment(name: "HandleEvent", baggage: context.baggage)
+        // TODO: record errors propagated in result types?
+        let decodedEvent = segment.subsegment(name: "DecodeIn") { _ in
+            self.decodeIn(buffer: event)
+        }
+        switch decodedEvent {
         case .failure(let error):
+            segment.addError(error)
+            segment.end()
             return context.eventLoop.makeFailedFuture(CodecError.requestDecoding(error))
         case .success(let `in`):
-            return self.handle(context: context, event: `in`).flatMapThrowing { out in
-                switch self.encodeOut(allocator: context.allocator, value: out) {
-                case .failure(let error):
-                    throw CodecError.responseEncoding(error)
-                case .success(let buffer):
-                    return buffer
+            // TODO: use NIO helpers?
+            let subsegment = segment.beginSubsegment(name: "HandleIn")
+            context.baggage = subsegment.baggage
+            return self.handle(context: context, event: `in`)
+                .always { result in
+                    if case .failure(let error) = result {
+                        subsegment.addError(error)
+                    }
+                    subsegment.end()
+                }
+                .flatMapThrowing { out in
+                    try context.tracer.segment(name: "EncodeOut", baggage: segment.baggage) { _ in
```

Comment:
Naming nitpick: I really would like to get all libs to be consistent with the use of the `context` parameter name, note that it can then accept a carrier and it becomes easier to just pass the `context: segment` / `context: span` / `context: context (the lambda context); the only exception is `context.baggage` (like in `Lambda.Context`)

Reply:
this will be "context" when using `TracingInstrument` API:

```swift
var span = tracer.startSpan(named: "EncodeOut", context: segment.baggage) 
```

I dont have strong opinion on that, but to avoid confusion in XRaySDK I use "context" for `XRayContext` type (which does have the X-Ray trace context, strongly typed) and "baggage" for `BaggageContext` which *may* have the X-Ray trace context

Reply:
What would you say about the following spelling though:

```
var context = BaggageContext()
context.xRay =  XRayContext()
// since also:
// context.zipkin = ZipkinState() // I call it state in some things I worked on,
//
// because it aligns with  https://www.w3.org/TR/trace-context/#combined-header-value
// "tracestate" where each tracer may carry their own state by a vendor identified key
//
// Example: vendorname1=opaqueValue1,vendorname2=opaqueValue2
```

also because one can use the BaggageContextCarrier to assign _through_ into the underlying baggage context;
This way all use sites, regardless if a carrier, raw baggage context, or any "my specific framework type" can use the same call site style.

Reply:
I made a ticket to discuss naming once more: https://github.com/slashmo/gsoc-swift-baggage-context/issues/23